modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionServiceImpl.java

```diff
@@ -555,9 +565,16 @@ private QueryPlan prepareFragment(BaseQueryContext ctx, String jsonFragment) {
                             fragment.serialized(),
                             ectx.topologyVersion(),
                             fragmentDesc,
-                            qry.parameters());
+                            fragmentsPerNode.get(nodeId).intValue(),
+                            qry.parameters(),
+                            parametersMarshalled
+                        );
 
                         messageService().send(nodeId, req);
+
+                        // Avoid marshaling of the same parameters for other nodes.
+                        if (parametersMarshalled == null)
```

Comment:
i don`t like this intricate code, but if you still want to realize it seems you can speedup a bit, check - after `messageService().send(nodeId, req);` you already have marshalled params in QueryStartRequest.paramsBytes, simply invoke it and reuse.